Infrastructure: Sentry integration

NicolasKeita opened the pull request:
#### Brief overview of PR changes/additions

You can view the demonstration video [here](https://epitechfr-my.sharepoint.com/:v:/g/personal/nicolas_keita_epitech_eu/IQCrhbKrGhS1QKjkcJffZdt4AZe5W1CoAGupA0eXmygsK2k?nav=eyJyZWZlcnJhbEluZm8iOnsicmVmZXJyYWxBcHAiOiJPbmVEcml2ZUZvckJ1c2luZXNzIiwicmVmZXJyYWxBcHBQbGF0Zm9ybSI6IldlYiIsInJlZmVycmFsTW9kZSI6InZpZXciLCJyZWZlcnJhbFZpZXciOiJNeUZpbGVzTGlua0NvcHkifX0&e=u4lG2f).

1) Added two compile-time variables:
    - `SENTRY_SEND_DEBUG`
    - `WITH_SENTRY`
```
cmake -S ./ -B build/ -DWITH_SENTRY=ON -DSENTRY_SEND_DEBUG=1
qmake6 ../src/mudlet.pro CONFIG+=WITH_SENTRY SENTRY_SEND_DEBUG=1
```

All PR code is wrapped with `IF (WITH_SENTRY)` blocks, making it easy to enable or disable the Sentry integration.

The `SENTRY_SEND_DEBUG `option is used to automatically upload debug files to Sentry.io. These files are required by Sentry to **display the exact line** of a crash.

Sentry requires both the executable and its corresponding debug symbols. Debug symbol files can be large (up to 350 MB for Mudlet), which is why this option ```SENTRY_SEND_DEBUG``` exists—to avoid uploading the full debug files on every CI push.
- Windows: executable ~~+ `.pdb` file~~ (all debug symbols are embedded directly in the Mudlet binary)
- Linux: executable + `.debug` file
- macOS: executable + `.dSYM` folder

Ideally, the executable is "stripped" of all debug symbols to reduce its size, while the debug symbol files contain all the necessary debugging information. Both the stripped executable and the debug files are sent to Sentry, which is sufficient for accurate crash reporting.

2) Added a secondary program: A small window prompts the user for consent to send the crash report. This program is located at `src/crash_reporter/main.cpp`.

3) Copied a mandatory Sentry-provided executable (crash_handler) required for proper functioning, which must be placed alongside the Mudlet executable.

4) Generated a mandatory cache directory at runtime for storing Sentry crash reports.
```
      Expected cache locations:
        Linux   : ~/.cache/mudlet/sentry
        macOS   : ~/Library/Caches/mudlet/sentry
        Windows : C:\Users\...\AppData\Local\Cache\Mudlet\sentry
```
5) Added a git submodule :  [sentry-native ](https://github.com/getsentry/sentry-native)

6) Added two CI secret variables
- `${{ secrets.SENTRY_DSN }}` sets the default DSN at compile time (developers can override it via an environment variable named `SENTRY_DSN`).
- `${{secrets.SENTRY_AUTH_TOKEN}}` is used to send debug files. (dev can also override it via a environment variable `SENTRY_AUTH_TOKEN`

#### How to Test

1. Create a Sentry account on the Sentry website.
2. Create a new project (name it `mudlet`) and obtain the project's DSN then add it to the console environment: export SENTRY_DSN="..."
3. `export MUDLET_CRASH_TEST=1` (thanks to this variable, Mudlet will crash whenever you open a profile. Let's use Sentry to see exactly where I inserted the crash.)
4. Start Mudlet like a normal user and open a game. Mudlet should crash.
5. You should now see the crash on the Sentry website.

But without the debug files, it doesn't show exactly where the crash is.

6. Obtain your organization’s auth token from the Sentry website and set it in your console environment:
   ```export=SENTRY_AUTH_TOKEN="..."```
7. Upload the debug files by recompiling with:
    - ```-DSENTRY_SEND_DEBUG=1```
       - Or the provided script:
    - ```CI/send_debug_files_to_sentry.sh```
8. Run Mudlet and trigger the crash.
9. Verify the crash details on the Sentry website.

#### Motivation for adding to Mudlet

/claim #7126,
fixes #7126.

#### Other info (issues closed, discussion etc)

I pushed a working snippet for qmake (mudlet.pro) aswell,
Although QMake is expected to be removed out soon, I added this snippet to ensure compatibility for the time being.

NicolasKeita committed:
Add: Sentry Integration

NicolasKeita committed:
Fix CI checks

NicolasKeita committed:
Fix: Strip windows executable

NicolasKeita committed:
Fix: Strip for windows in the right order

vadi2 commented on src/cmake/WithSentry.cmake:
```diff
@@ -0,0 +1,162 @@
+        COMMENT "Creating separate debug file and stripping executable"
+    )
+else()
+    add_custom_command(TARGET ${EXE_MUDLET_TARGET} POST_BUILD
+        COMMAND strip --strip-debug $<TARGET_FILE:${EXE_MUDLET_TARGET}>
+    )
```
We don't use MSVC on Windows.

vadi2 commented on .github/workflows/build-mudlet-win.yml:
```diff
@@ -32,13 +32,28 @@ jobs:
+      shell: msys2 {0}
+      run: |
+        echo "WITH_SENTRY=yes" >> $GITHUB_ENV
+        if [ -n "$SENTRY_SEND_DEBUG_SECRET" ]; then
```
Will also trigger on builds in branches made in this repository by a core dev member? Maybe better to limit it to development, master, and release-* branches.

NicolasKeita replied:
I updated the way I set the environment variables:

```diff
    - name: (Windows) Build
      shell: msys2 {0}
      env:
        GITHUB_REPO_TAG: ${{ github.event_name == 'push' && startsWith(github.ref, 'refs/tags/') }}
        GITHUB_SCHEDULED_BUILD: ${{ github.event_name == 'schedule' || github.event.inputs.scheduled == 'true' }}
        GITHUB_PULL_REQUEST_NUMBER: ${{ github.event.pull_request.number }}
        GITHUB_PULL_REQUEST_HEAD_SHA: ${{ github.event.pull_request.head.sha }}
+       SENTRY_AUTH_TOKEN: ${{ secrets.SENTRY_AUTH_TOKEN }}
+       SENTRY_SEND_DEBUG: ${{ startsWith(github.ref, 'refs/tags/') && '1' || '0' }}
+       SENTRY_DSN: ${{ secrets.SENTRY_DSN }}
+       WITH_SENTRY: "ON"
      run: $GITHUB_WORKSPACE/CI/build-mudlet-for-windows.sh
```
      
By putting them here, I don’t think it will impact other branches, since at the start of the file there is:
    


   push:
    branches: [master, development, release-*]
```

vadi2 commented on CI/setup-windows-sdk.sh:
```diff
@@ -125,7 +125,11 @@ while true; do
     "mingw-w64-${BUILDCOMPONENT}-meson" \
     "mingw-w64-${BUILDCOMPONENT}-ninja" \
     "mingw-w64-${BUILDCOMPONENT}-assimp" \
-    "mingw-w64-${BUILDCOMPONENT}-jq"; then
```
We might want to keep jq

NicolasKeita replied:
That was unintentional, I’ve added it back.

```diff
+    "mingw-w64-${BUILDCOMPONENT}-jq"; then

vadi2 commented on src/crash_reporter/CMakeLists.txt:
```diff
@@ -0,0 +1,71 @@
+project(MudletCrashReporter LANGUAGES CXX)
+
+set(CMAKE_CXX_STANDARD 17)
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
```
Lets align this with the main CMakeLists options (C++20 and the minimum cmake version)

NicolasKeita replied:
Updated it to : 

```diff

-cmake_minimum_required(VERSION 3.16)

-set(CMAKE_CXX_STANDARD 17)
-set(CMAKE_CXX_STANDARD_REQUIRED ON)

+cmake_minimum_required(VERSION 3.25.1)

+set(CMAKE_CXX_STANDARD 20)
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
+set(CMAKE_CXX_EXTENSIONS OFF)
```

vadi2 replied:
Solid.

vadi2 commented on src/crash_reporter/main.cpp:
```diff
@@ -0,0 +1,61 @@
+
+    box.setWindowTitle("Mudlet Crash");
+    box.setText("<div align='center'>Mudlet crashed.<br>Would you like to send a crash report?</div>");
+    box.setStandardButtons(QMessageBox::Yes | QMessageBox::No);
```
Really nice. Could you make this i18n friendly and add an option, selected by default, to send + remember sending in the future?

NicolasKeita replied:
I updated the PR,  wrapped the strings with QCoreApplication::translate() and added an option to remember sending in the future :
https://github.com/Mudlet/Mudlet/pull/8540#issuecomment-3568203742

vadi2 commented on src/mudlet-lua/lua/DateTime.lua:
```diff
@@ -46,6 +46,32 @@ datetime = {
   }
 }
 
+-- rex.gsub was crashing so here is a homemade
```
Very interesting, got more details on this?

NicolasKeita committed:
Improve, Taking in consideration coderabbitai

NicolasKeita committed:
Improve, Taking in consideration coderabbitai

NicolasKeita committed:
Improve, Taking in consideration coderabbitai

NicolasKeita committed:
Improve, Taking in consideration coderabbitai

NicolasKeita committed:
Improve, Taking in consideration coderabbitai

NicolasKeita committed:
Improve, Taking in consideration coderabbitai

NicolasKeita commented on CI/package-mudlet-for-windows.sh:
```diff
@@ -101,11 +101,21 @@ if [ ! -f "${GITHUB_WORKSPACE_UNIX_PATH}/build-${MSYSTEM}/${BUILD_CONFIG}/mudlet
+
+if [ "${WITH_SENTRY}" = "yes" ]; then
+    FILES_TO_COPY+=("crashpad_handler.exe" "crashpad_wer.dll")
 fi
```
In our case, we actually want the CI to fail if the Sentry binaries don’t exist, because these two binaries are mandatory.

The suggested code would only hide this issue.

NicolasKeita commented on src/cmake/WithSentry.cmake:
```diff
@@ -0,0 +1,162 @@
+add_dependencies(${LIB_MUDLET_TARGET} sentry_native)
+
+file(READ "${CMAKE_SOURCE_DIR}/src/sentry_dsn.txt" SENTRY_DSN_RAW)
+string(STRIP "${SENTRY_DSN_RAW}" SENTRY_DSN)
```
In our case, we actually want the CI to fail if sentry_dsn.txt doesn’t exist, because this file is mandatory.

The suggested code would only hide this issue.

NicolasKeita commented on src/cmake/WithSentry.cmake:
```diff
@@ -0,0 +1,142 @@
+    endif()
+
+    list(APPEND SENTRY_CMAKE_ARGS
+        "-DCMAKE_OSX_ARCHITECTURES=${ARCH_LIST}"
+        "-DCMAKE_OSX_SYSROOT=${MACOSX_SYSROOT}"
+    )
+endif()
+include(ExternalProject)
+ExternalProject_Add(
+    sentry_native
+    SOURCE_DIR ${CMAKE_SOURCE_DIR}/3rdparty/sentry-native
+    CMAKE_ARGS ${SENTRY_CMAKE_ARGS}
+)
+
+add_dependencies(${LIB_MUDLET_TARGET} sentry_native)
+
+file(READ "${CMAKE_SOURCE_DIR}/src/sentry_dsn.txt" SENTRY_DSN_RAW)
+string(STRIP "${SENTRY_DSN_RAW}" SENTRY_DSN)
+
+target_compile_options(${LIB_MUDLET_TARGET} PRIVATE -g)
+
+if(WIN32)
+    target_link_options(${EXE_MUDLET_TARGET} PRIVATE -g -gcodeview)
+    string(REPLACE "-Wl,-s" "" CMAKE_EXE_LINKER_FLAGS_RELEASE "${CMAKE_EXE_LINKER_FLAGS_RELEASE}")
+    string(REPLACE "-Wl,-s" "" CMAKE_SHARED_LINKER_FLAGS_RELEASE "${CMAKE_SHARED_LINKER_FLAGS_RELEASE}")
+endif()
+
+target_compile_definitions(${LIB_MUDLET_TARGET} PUBLIC
+    WITH_SENTRY
+    SENTRY_DSN="${SENTRY_DSN}"
+    APP_DIR_PATH="${CMAKE_BINARY_DIR}/src"
+    SENTRY_BUILD_STATIC
+)
+target_compile_definitions(${EXE_MUDLET_TARGET} PUBLIC
+    WITH_SENTRY
+    SENTRY_DSN="${SENTRY_DSN}"
+    APP_DIR_PATH="${CMAKE_BINARY_DIR}/src"
+    SENTRY_BUILD_STATIC
+)
+
```
- APP_DIR_PATH baked into runtime—fails after installation
    - as been fixed.
-  Hard-coded toolchain (clang/clang++/Ninja) mismatches main project configuration
    - We want to compile sentry-native with clang, so I added the package to the CI to ensure the required toolchain is available.
 - cURL is present on Windows by default, and Bash is available in the MSYS2 environment, which is the recommended environment for developing Mudlet on Windows.

NicolasKeita commented on src/crash_reporter/main.cpp:
```diff
@@ -0,0 +1,61 @@
+        return 1;
+    }
+
+    QApplication    app(argc, argv);
+    QMessageBox     box;
+
+    box.setWindowTitle("Mudlet Crash");
+    box.setText("<div align='center'>Mudlet crashed.<br>Would you like to send a crash report?</div>");
+    box.setStandardButtons(QMessageBox::Yes | QMessageBox::No);
+    if (box.exec() == QMessageBox::Yes) {
+        sendCrashReport(argv[1]);
+    }
+    return 0;
+}
+
+void sendCrashReport(const char *envelopePath)
+{
+    sentry_options_t *  opts = sentry_options_new();
+    sentry_envelope_t * envelope = sentry_envelope_read_from_file(envelopePath);
+
+    sentry_options_set_dsn(opts, SENTRY_DSN);
+    sentry__transport_startup(opts->transport, opts);
+    sentry__transport_send_envelope(opts->transport, envelope);
+    sentry__transport_shutdown(opts->transport, 30 * 1000);
+}
```
1) 

> This will break if the Sentry API changes.

The Sentry submodule is pinned to a specific commit. The API will not change.
2) has been fixed.
3) The program is short-lived by design. There is no impact on users.
4) Sentry does it for us. It returns NULL if that is the case.

NicolasKeita committed:
Remove previous comments

NicolasKeita commented on src/SentryWrapper.cpp:
```diff
@@ -0,0 +1,67 @@
+ *   macOS   : ~/Library/Caches/mudlet/sentry
+ *   Windows : C:\Users\...\AppData\Local\Cache\Mudlet\sentry
+ */
+void initSentry()
+{
+    #ifdef WITH_SENTRY
+        sentry_options_t* options = sentry_options_new();
+
+        if (!options) {
+            return;
+        }
+        sentry_options_set_dsn(options, SENTRY_DSN);
+        QString path = QStandardPaths::writableLocation(QStandardPaths::CacheLocation) + "/mudlet/sentry";
+        sentry_options_set_database_path(options, path.toUtf8().constData());
+        sentry_options_set_release(options, "mudlet@" APP_VERSION);
+        sentry_options_set_handler_path(options, makeExecutablePath(APP_DIR_PATH, "crashpad_handler").c_str());
+        sentry_options_set_external_crash_reporter_path(options, makeExecutablePath(APP_DIR_PATH, "MudletCrashReporter").c_str());
+
+        sentry_init(options);
+        auto sentryClose = qScopeGuard([] { sentry_close(); });
+    #endif
+}
+
+std::string makeExecutablePath(const std::string& dir, const std::string& name) {
+    #ifdef _WIN32
+        return dir + "/" + name + ".exe";
+    #else
+        return dir + "/" + name;
+    #endif
+}
```
1 2 has been fixed.
3) QStandardPaths::CacheLocation only adds the app-specific subdirectory if queried after the application is created. Since we call it before the app exists, it does not yet know the app name (“mudlet”).

vadi2 commented:
@NicolasKeita feel free to stop when you feel coderabbit comments are enough, your call

NicolasKeita commented:
@vadi2 I’ve just finished reviewing all of the CodeRabbit comments. The PR should now be in good shape. I had actually missed one thing initially, so codeRabbit was helpful :)

NicolasKeita committed:
Add a way to test Sentry more easily

NicolasKeita committed:
Removing sentry_dsn.txt

vadi2 commented:
A few notes for the original post. When compiling with sentry enabled,

* git submoduled need to be checked out recursively
* SENTRY_AUTH_TOKEN needs to be set
* profile needs to be loaded in online mode, not offline
* sanitizers need to be disabled for sentry reporting to work

vadi2 commented on src/mudlet-lua/lua/StringUtils.lua:
```diff
@@ -215,3 +215,24 @@ function f(supersecretstringvariablenocollision)
   end))
 end
+
+-- Homemade rex.gsub(), original was crashing.
```
What was the issue with this?

NicolasKeita replied:
This crash occurs on any version of Mudlet on Linux when linked with CURL — that alone is enough to trigger it:

```
target_link_libraries(
    ${LIB_MUDLET_TARGET}
    ...
    curl
)
```

Then the Lua tests crash:
```
export AUTORUN_BUSTED_TESTS=1
export TESTS_DIRECTORY=src/mudlet-lua/tests
xvfb-run --auto-servernum ./build/src/mudlet --profile "Mudlet self-test" --mirror
```

You get a “bad free” / “invalid pointer” crash:
```
SUMMARY: AddressSanitizer: bad-free
==21764==ERROR: AddressSanitizer: attempting free on address which was not malloc()-ed: 0x7fff89459b80 in thread T0
    #0 0x71db098fc4d8 in free ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:52
    #1 0x71dafbbafe2b in freelist_free (/usr/local/lib/lua/5.1/rex_pcre.so+0x3e2b)
    #2 0x71dafbbb2c08 in algf_gsub (/usr/local/lib/lua/5.1/rex_pcre.so+0x6c08)
```

I think, the reason is that rex.gsub() internally allocates a temporary linked list of nodes to do its job. Then, At the end, it frees the list with freelist_free().

I asked ChatGPT to recreate rex.gsub() using other rex_pcre functions but without freeing.

ChatGPT explains that it might be because curl or one of its dependencies redefines malloc/free.

vadi2 replied:
Could I interest you in implementing a custom transport using Qt functionality instead to see if that'll work better, and not use curl? https://docs.sentry.io/platforms/native/configuration/transports

Maybe our tests found one problem, but out test coverage is not extensive, and we don't know what else could break.

NicolasKeita replied:
I’ll take a closer look at it.

vadi2 replied:
Double-checking, is it necessary for curl to be linked to Mudlet? Isn't the new binary doing the uploading?

NicolasKeita replied:
You're right — I moved the cURL dependency into the new crash-reporter binary.

The main Mudlet executable now links against a transport-less Sentry.  

```diff
- -DSENTRY_TRANSPORT=curl
+ -DSENTRY_TRANSPORT=none
```
(Building Sentry with `-DSENTRY_TRANSPORT=curl`,  forces the final executable to link against cURL as well.)

By keeping cURL out of the main executable, the allocator/memory conflict disappears entirely.  
I was therefore able to revert `rex.gsub()` to its original implementation — no Lua changes are needed anymore, and all tests pass again.

vadi2 reviewed (commented):
Confirmed works well on Linux. Could you resolve the merge conflicts? The windows build has expired since, and I'd like to test it there.

NicolasKeita committed:
Style cleanup

NicolasKeita committed:
Style cleanup

NicolasKeita committed:
Add QComboBox for crash report policy with persistence via QSettings

NicolasKeita commented:
I updated the PR. Solved the merge conflict.

> Could you make this i18n friendly and add an option, selected by default, to send + remember sending in the future?

Here’s what I came up with:

<img width="449" height="126" alt="image" src="https://github.com/user-attachments/assets/cc580b6d-aafb-4b28-ba5d-1641da2aba5a" />

There’s also a QComboBox in Preferences, in case the user wants to change it later:

<img width="1026" height="624" alt="image" src="https://github.com/user-attachments/assets/536471fb-d6fd-4f7c-9afb-37284396632c" />

vadi2 commented:
Looks solid, thanks.

NicolasKeita committed:
Remove libcurl from Mudlet main app

NicolasKeita committed:
only strip executable in Release mode

vadi2 commented:
/refresh links

vadi2 commented:
How can I get the crash handler to trigger on Windows?


https://github.com/user-attachments/assets/fe29c49d-bfc1-4ae7-b7e2-853aea691306


I previously got it working macOS and Linux in [this PR](https://github.com/Mudlet/Mudlet/pull/7664), but Windows was the hard part.

NicolasKeita committed:
Add crash handler to the final package, on Windows

NicolasKeita committed:
Add crash handler executable to the final package

vadi2 commented:
Clang-tidy should work with latest development merged in

NicolasKeita committed:
Add: Sentry executable to the Linux/MacOS final bundle

NicolasKeita commented:
> How can I get the crash handler to trigger on Windows?
> 
>  Screencast.from.2025-11-28.06-33-55.mp4 
> I previously got it working macOS and Linux in [this PR](https://github.com/Mudlet/Mudlet/pull/7664), but Windows was the hard part.



Apologies, I forgot to include `crashpad_handler.exe` and `MudletCrashReporter.exe` in the final Windows package. I've just pushed the necessary changes.

```diff
+ if [ "${WITH_SENTRY}" = "ON" ]; then
+    FILES_TO_COPY+=("crashpad_handler.exe" "crashpad_wer.dll" "MudletCrashReporter.exe")
+ fi
```

Here's a screencast showing it in action on Windows: [link](https://epitechfr-my.sharepoint.com/:v:/g/personal/nicolas_keita_epitech_eu/IQB_4eanCTz8QYz9XxhmXJ5OASFuGTjzsd0eHkBNn0cPocs)

--------------------------------------------------------

Also, in case you're wondering about the failing checks in your PR on `ubuntu / gcc / lua tests`:

```
Not setup to auto run display benchmark.
Setup to auto-run tests - commencing.
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++free(): invalid pointer
```

This was the crash I mentioned related to rex.gsub().

--------------------------------------------------------

If you want to go further and see the exact line where a crash occurs in Sentry.io, we need to send two files with debug information

```
mudlet.exe (including debug symbols, ~350 MB)
mudlet.pdb
```

The final executables that are distributed do not contain debug symbols, so they are much smaller and lightweight.

There are a few ways to send the debug symbols **in between compilation and the packaging of the final executables without debug symbols:**

1) Automatically when creating a release
Debug files are automatically uploaded when creating a new release.
```
SENTRY_SEND_DEBUG: ${{ startsWith(github.ref, 'refs/tags/') && '1' || '0' }}
```

2) Automatically during compilation
Compile with `-DSENTRY_SEND_DEBUG=1` to send debug files during the build:
```
cmake -S ./ -B build/ -DWITH_SENTRY=ON -DSENTRY_SEND_DEBUG=1 -DCMAKE_BUILD_TYPE=Debug
cmake --build build/ -j4
```

3) Manually
Compile with WITH_SENTRY=ON, then upload the debug files using either:

- My script  `./CI/send_debug_files_to_sentry.sh build/src/mudlet.exe` 
, or
- The official [Sentry CLI](https://docs.sentry.io/cli/installation/) `sentry-cli debug-files upload -p <project> /path/to/files...`

vadi2 commented:
Thanks! I got the dialog to come up and it did report the crash. No details available, but that's because as you mention uploading is limited to certain scenarios.

Thinking about it, we might as well upload all debug info whenever sentry itself is enabled, right? It wouldn't cause the executable size that players get in testing PRs, PTBs or releases to be larger because that information is stripped after the upload to sentry.

vadi2 committed:
Improve crash reporter dialog UX

- Replace confusing options with clearer choices:
  "Send this time" / "Always send" / "Don't send"
- Only save preference when "Always send" is chosen
- Make "Send this time" the default (least committal)
- Add spacing between dialog text lines

NicolasKeita commented:
@vadi2 

In my current design, Sentry activation and debug-file uploading are two separate options.

I created this separation in case we don’t want to upload ~300 MB on every CI workflow (currently Sentry doesn’t impose any storage limit, but I don’t think that will stay that way forever).

That said, you’re right—the debug info is stripped from the final executable, so its size for players won’t be affected whether we upload the debug files or not.

Current workflow design:

- `WITH_SENTRY="ON"` is _always_ enabled:

```
on:
  push:
    branches: [master, development, release-*]  # Pushes to master/dev/release-*
    tags: [Mudlet-*]                            # Pushes of Mudlet-* tags
  pull_request:                                 # Any pull request
  workflow_dispatch:                            # Manual workflow triggers via GitHub
```

- `SENTRY_SEND_DEBUG=1` happens only for release tags:

```
env:        
         SENTRY_SEND_DEBUG: ${{ startsWith(github.ref, 'refs/tags/') && '1' || '0' }}
run: 
          $GITHUB_WORKSPACE/CI/build-mudlet-for-windows.sh
```

To experiment (temporarily or not), knowing that debug files will be uploaded on every CI workflow, we can slightly adjust the design:

Enabling debug-file upload on every CI workflow:
```diff
env:
        SENTRY_AUTH_TOKEN: ${{ secrets.SENTRY_AUTH_TOKEN }}
        SENTRY_DSN: ${{ secrets.SENTRY_DSN }}
        WITH_SENTRY: "ON"
+     SENTRY_SEND_DEBUG: '1'
-     SENTRY_SEND_DEBUG: ${{ startsWith(github.ref, 'refs/tags/') && '1' || '0' }}
run:
      $GITHUB_WORKSPACE/CI/build-mudlet-for-windows.sh
```

Feel free to let me know if we need to adjust the design, such as when to compile with WITH_SENTRY or when to upload the debug files.

NicolasKeita committed:
Merge pull request #2 from Mudlet/feature/crash-reporter-ux

Improve crash reporter dialog UX

vadi2 commented:
Fair call on the size. Lets try uploading them on release and PTB builds only, and as a once-off on this PR (so the stack unwinding can be tested)

NicolasKeita committed:
Add: Send Sentry debug files on PTB builds

NicolasKeita committed:
Fix: Add missing parentheses in SENTRY_SEND_DEBUG expression

NicolasKeita committed:
Add: Always sending debug files

NicolasKeita committed:
Verify presence of SENTRY_AUTH_TOKEN in CI

NicolasKeita committed:
Verify presence of SENTRY_AUTH_TOKEN in CI

NicolasKeita committed:
Verify presence of SENTRY_AUTH_TOKEN in CI

NicolasKeita committed:
Verify presence of SENTRY_AUTH_TOKEN in CI

NicolasKeita committed:
Verify presence of SENTRY_AUTH_TOKEN in CI

NicolasKeita committed:
Fix using normal variable instead of secret variable for SENTRY_AUTH_TOKEN

NicolasKeita committed:
Restarting CI

NicolasKeita committed:
Verify presence of SENTRY_AUTH_TOKEN in CI

NicolasKeita committed:
Verify presence of SENTRY_AUTH_TOKEN in CI

NicolasKeita committed:
Verify presence of SENTRY_AUTH_TOKEN in CI

NicolasKeita committed:
Hard coding SENTRY_DSN and SENTRY_AUTH_TOKEN

NicolasKeita committed:
Adding the correct Sentry Auth Token

vadi2 commented:
Works great - 
<img width="1095" height="246" alt="image" src="https://github.com/user-attachments/assets/114ac2d5-ab35-4314-8f9a-796e5ef32143" />

NicolasKeita committed:
Using secrets for variable SENTRY_AUTH_TOKEN and SENTRY_DSN

SlySven commented:
> Sentry requires both the executable and its corresponding debug symbols. Debug symbol files can be large (up to 350 MB for Mudlet), which is why this option SENTRY_SEND_DEBUG exists—to avoid uploading the full debug files on every CI push.
>
>    Windows: executable + .pdb file
>    Linux: executable + .debug file
>    macOS: executable + .dSYM folder

This is RONG - Windows builds use the MSYS2+Mingw64 toolchain and `clang` and `gcc` BOTH produce `.debug` files on that platform - `.pdb` files are MSVC files and we don't make those.

SlySven commented on src/mudlet.pro:
```diff
@@ -613,6 +613,65 @@ contains( DEFINES, INCLUDE_UPDATER ) {
+    !exists($$SENTRY_PATH/install) {
+        message("Sentry install missing, building sentry-native from sources")
+
+        system(cmake -S $$SENTRY_PATH -B $$SENTRY_PATH/build -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ -DSENTRY_BACKEND=crashpad -D SENTRY_BUILD_SHARED_LIBS=OFF --log-level=ERROR)
```
This trivial error is still in place and needs to be applied should anyone try to use QMake. Yes I know it is deprecated but in some situations it is still usable/useful (particularly given how hard CMake makes it to examine files not being directly used in the compilation and linkage stages of the build process)...

SlySven commented on src/mudlet.pro:
```diff
@@ -613,6 +613,65 @@ contains( DEFINES, INCLUDE_UPDATER ) {
+
+    SENTRY_PATH = $$PWD/../3rdparty/sentry-native
+
+    !exists($$SENTRY_PATH/install) {
+        message("Sentry install missing, building sentry-native from sources")
+
+        system(cmake -S $$SENTRY_PATH -B $$SENTRY_PATH/build -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ -DSENTRY_BACKEND=crashpad -D SENTRY_BUILD_SHARED_LIBS=OFF --log-level=ERROR)
+        system(cmake --build $$SENTRY_PATH/build --parallel)
+        system(cmake --install $$SENTRY_PATH/build --prefix $$SENTRY_PATH/install)
+    }
```
This is still unresolved. (But like the prior issue I am looking at it right now as I try to get `development` branch code buildable in my development environment - currently Windows but am thinking about other OSes as well).

SlySven commented on src/mudlet.pro:
```diff
@@ -613,6 +613,65 @@ contains( DEFINES, INCLUDE_UPDATER ) {
+
+
+    QMAKE_POST_LINK += $$quote($$QMAKE_MOVE $$OUT_PWD/mudlet.pdb $$APP_DIR_PATH/) ;
+    QMAKE_POST_LINK += $$quote(cp -f $$SENTRY_PATH/install/bin/* $$APP_DIR_PATH/) ;
```
`QMAKE_MOVE` now present or the `QMAKE_COPY` imagined here are both AI fantasies - there is not such function in QMake. :rage:

NicolasKeita commented:
@SlySven 

I agree that by default, GCC under MinGW generates executables containing DWARF debug information, not CodeView.

_DWARF symbols can be split into a separate `.debug` file, whereas the equivalent external symbol file for CodeView would be a `.pdb` file._
Therefore, with our current toolchain, no .pdb files are generated by default.

Additionally, Sentry specifically requires debug information in CodeView format on Windows in order to correctly process stack traces.

For this reason, I enabled the GCC option -[gcodeview ](https://gcc.gnu.org/onlinedocs/gcc/Debugging-Options.html)(which I have read somewhere is still a little bit experimental) :

`target_link_options(${EXE_MUDLET_TARGET} PRIVATE -g -gcodeview)`

With this option enabled, the final mudlet.exe contains the CodeView symbols itself, and uploading the executable is enough for Sentry to resolve stack traces with accuracy. Sentry accepts this approach, so we do not need two separate files (binary + external debug file) for windows.


After uploading to Sentry, the executable is stripped to reduce its size.
Even in stripped form, Sentry is still able to correctly process crash information with accuracy.

### About the .pdb file:

Using `-gcodeview` also causes GCC to emit a .pdb file.
However because the option is still experimental, this .pdb does not contain the full set of symbols and is not reliable.

For this reason, the .pdb file is simply removed after the build and not used.
```
    add_custom_command(TARGET ${EXE_MUDLET_TARGET} POST_BUILD
        COMMAND ${CMAKE_COMMAND} -E remove "${CMAKE_BINARY_DIR}/mudlet.pdb"
    )
```

My original post was confusing when mentioning an external .pdb file. I apologize for that and have updated it.

------------------------

On another note, I think Sentry could benefit from supporting DWARF symbols (`.debug` files) for Windows executables, as MinGW is a commonly used toolchain.

SlySven commented:
> On another note, I think Sentry could benefit from supporting DWARF symbols (.debug files) for Windows executables, as MinGW is a commonly used toolchain.

Let's see if https://github.com/getsentry/sentry/issues/104738 gets any traction...

SlySven commented:
> I agree that by default, GCC under MinGW generates executables containing DWARF debug information, not CodeView.
>
> DWARF symbols can be split into a separate .debug file, whereas the equivalent external symbol file for CodeView would be a .pdb file.
>Therefore, with our current toolchain, no .pdb files are generated by default.
>
>Additionally, Sentry specifically requires debug information in CodeView format on Windows in order to correctly process stack traces.
>
>For this reason, I enabled the GCC option -[gcodeview ](https://gcc.gnu.org/onlinedocs/gcc/Debugging-Options.html)(which I have read somewhere is still a little bit experimental) :

I understand that `clang` does a better job and I have tried to get things to build with the `MINGW64` environments `clang` but still using CMake {as the QMake build file was *unparsable* by Qmake} - I've fixed things enough that not setting `WITH_SENTRY=ON` will build and also completely excludes the `./src/SentryWrapper.cpp` and `./src/SentryWrapper.h` files from the build. This is so otherwise the build borks on the absence of sentry related files when they are not needed/present which ~~hackers~~ developers on their own builds won't need. However I am running into a problem where two different parts of the `sentry-native` build use different versions of the same files.
Edited to hide some personal path details by making some absolute paths relative to base of Mudlet source code:
```
.\mudlet\3rdparty\sentry-native\external\crashpad\third_party\mpack\mpack.h:7256: error: multiple definition of `mpack_tree_size'; ../../3rdparty/sentry-native/libsentry.a(mpack.c.obj):.\mudlet\3rdparty\sentry-native\vendor/mpack.h:6230: first defined here
C:/msys64/mingw64/bin/ld: ../../3rdparty/sentry-native/crashpad_build/client/libcrashpad_client.a(crash_report_database.cc.obj): in function `mpack_tree_size':
./mudlet/3rdparty/sentry-native/external/crashpad/third_party/mpack/mpack.h:7256: multiple definition of `mpack_tree_size'; ../../3rdparty/sentry-native/libsentry.a(mpack.c.obj):.\mudlet\3rdparty\sentry-native\vendor/mpack.h:6230: first defined here
```

Basically, these (version **1.0**) files:
```
.\mudlet\3rdparty\sentry-native\vendor\mpack.h
.\mudlet\3rdparty\sentry-native\vendor\mpack.c
```
clash with these (version **1.1.1**) files:
```
.\mudlet\3rdparty\sentry-native\external\crashpad\third_party\mpack\mpack.h
.\mudlet\3rdparty\sentry-native\external\crashpad\third_party\mpack\mpack.cc
```

:shrug:

NicolasKeita commented:
I wasn’t able to reproduce the exact same issue on my side, but I did encounter similar 
“multiple definition” problems when building sentry-native multiple times with different 
configurations.

In my experience, these issues are usually resolved by starting from a completely clean state, for example:

```
  rm -rf 3rdparty/
  git restore 3rdparty/
  git submodule update --init --recursive
```

<details>
For details, this is how sentry-native is currently compiled via CMake:

```
set(SENTRY_COMMON_ARGS
    -DCMAKE_BUILD_TYPE=RelWithDebInfo
    -DCMAKE_C_COMPILER=clang
    -DCMAKE_CXX_COMPILER=clang++
    -DSENTRY_BACKEND=crashpad
    -DSENTRY_BUILD_SHARED_LIBS=OFF
    -DSENTRY_INTEGRATION_QT=OFF
    -G Ninja
)

ExternalProject_Add(
    sentry_without_transport
    SOURCE_DIR ${SENTRY_PATH}
    CMAKE_ARGS
        -DCMAKE_INSTALL_PREFIX=${SENTRY_PATH}/install_without_transport
        ${SENTRY_COMMON_ARGS}
        -DSENTRY_TRANSPORT=none
)
```

These correspond to the two manual commands:

```
cmake -S 3rdparty/sentry-native -B 3rdparty/sentry-native/build \
    -DCMAKE_INSTALL_PREFIX=3rdparty/sentry-native/install_without_transport \
    -DCMAKE_BUILD_TYPE=RelWithDebInfo \
    -DCMAKE_C_COMPILER=clang \
    -DCMAKE_CXX_COMPILER=clang++ \
    -DSENTRY_BACKEND=crashpad \
    -DSENTRY_BUILD_SHARED_LIBS=OFF \
    -DSENTRY_INTEGRATION_QT=OFF \
    -DSENTRY_TRANSPORT=none \
    -G Ninja

cmake --build 3rdparty/sentry-native/build --target install
```


Post-build, on my setup, `mpack_tree_size` is only present in `libsentry.a`. `libcrashpad_client.a` does not contain it:

```
  $ nm 3rdparty/sentry-native/install_with_transport/lib/libsentry.a | grep "mpack_tree_size"
  0000000000002ec0 T mpack_tree_size

  $ nm 3rdparty/sentry-native/install_with_transport/lib/libcrashpad_client.a | grep "mpack_tree_size"
  (no output)
```
</details>

Hopefully starting from a clean state helps avoid these conflicts, and in the meanwhile I fixed/updated the mudlet.pro in a [PR](#8640 ) to keep backward compatibility.

SlySven commented:
> In my experience, these issues are usually resolved by starting from a completely clean state, for example:
> ```
>  rm -rf 3rdparty/
>  git restore 3rdparty/
>  git submodule update --init --recursive
>  ```

Initially I thought that `rm -rf 3rdparty/` would be a problem for me as I thought it would wipe out all my local repositories of ALL the 3rdparty stuff - but realising that the git submodule data is actually stored under the parent (Mudlet) `.git` subdirectory.
